src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -307,44 +336,118 @@ private void EventLoop()
         {
             try
             {
-                bool shutdown = false;
-                while (!shutdown)
+                if (IsAio)
+                {
+                    EventLoopBodyAio();
+                }
+                else
                 {
-                    int numEvents = EventBufferCount;
-                    Interop.Error err = Interop.Sys.WaitForSocketEvents(_port, _buffer, &numEvents);
-                    if (err != Interop.Error.SUCCESS)
+                    EventLoopBody();
+                }
+
+                FreeNativeResources();
+            }
+            catch (Exception e)
+            {
+                Environment.FailFast("Exception thrown from SocketAsyncEngine event loop: " + e.ToString(), e);
+            }
+        }
+
+        private void EventLoopBody()
+        {
+            bool shutdown = false;
+            while (!shutdown)
+            {
+                int numEvents = EventBufferCount;
+                Interop.Error err = Interop.Sys.WaitForSocketEvents(_port, _buffer, &numEvents);
+                if (err != Interop.Error.SUCCESS)
+                {
+                    throw new InternalException(err);
+                }
+
+                // The native shim is responsible for ensuring this condition.
+                Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
+
+                for (int i = 0; i < numEvents; i++)
+                {
+                    IntPtr handle = _buffer[i].Data;
+                    if (handle == ShutdownHandle)
                     {
-                        throw new InternalException(err);
+                        shutdown = true;
```

Comment:
the shutdown is implemented in a very interesting way: the epoll does not support canceling, but afaik we have some pipe where we write cancel request, which is represented by a file descriptior that is monitored by epool itself. So if there is new data in it, epoll receives a notification and we stop here.